crates/matrix-sdk-ui/src/timeline/event_handler.rs

```diff
@@ -1150,13 +1155,27 @@ impl<'a, 'o> TimelineEventHandler<'a, 'o> {
                     None => self.meta.new_timeline_item(item),
                 };
 
-                // Keep push semantics, if we're inserting at the front or the back.
-                if insert_idx == self.items.len() {
-                    self.items.push_back(new_item);
-                } else if insert_idx == 0 {
+                // We are about to insert the `new_item`, great! Though, we try to keep a
+                // precise insertion semantics here:
+                //
+                // * _push front_:
+                //   * if `must_push_front`, i.e. if the position is a `TimelinePosition::Start`
+                //   * if the index is 0 _but_ the items aren't empty
+                //
+                // * _push back_:
+                //   * if the index represents the end of items
+                //
+                // * _insert at_:
+                //   * for any other index value
+                if must_push_front || (insert_index == 0 && self.items.is_empty().not()) {
+                    trace!("Adding new remote timeline item at the front");
                     self.items.push_front(new_item);
+                } else if insert_index == self.items.len() {
```

Comment:
The order was very deliberate before, with the assumption that new events coming from sync will happen more frequently that prepending (because loading from cache). Can you revert and put the `if insert_index == self.items.len()` check first, please?

Reply:
It becomes:

```rust
if !must_push_front && (insert_index == self.items.len()) {
    self.items.push_back(new_item);
} else if must_push_front || insert_index == 0 {
    self.items.push_front(new_item);
} else {
    self.items.insert(insert_index, new_item);
}
```